example/restricted-boltzmann-machine/binary_rbm_gluon.py

```diff
@@ -0,0 +1,139 @@
+parser.set_defaults(cuda=True)
+
+args = parser.parse_args()
+print(args)
+
+### Global environment
+
+mx.random.seed(pyrnd.getrandbits(32))
+ctx = mx.gpu(args.device_id) if args.cuda else mx.cpu()
+
+
+### Prepare data
+
+def data_transform(data, label):
+    return data.astype(np.float32) / 255, label.astype(np.float32)
+
+mnist_train_dataset = mx.gluon.data.vision.MNIST(train=True, transform=data_transform)
+mnist_test_dataset = mx.gluon.data.vision.MNIST(train=False, transform=data_transform)
+img_height = mnist_train_dataset[0][0].shape[0]
+img_width = mnist_train_dataset[0][0].shape[1]
+num_visible = img_width * img_height
+
+# This generates arrays with shape (batch_size, height = 28, width = 28, num_channel = 1)
+train_data = mx.gluon.data.DataLoader(mnist_train_dataset, args.batch_size, shuffle=True, num_workers=args.data_loader_num_worker)
+test_data = mx.gluon.data.DataLoader(mnist_test_dataset, args.batch_size, shuffle=True, num_workers=args.data_loader_num_worker)
+
+### Train
+
+rbm = BinaryRBMBlock(num_hidden=args.num_hidden, k=args.k, for_training=True, prefix='rbm_')
+rbm.initialize(mx.init.Normal(sigma=.01), ctx=ctx)
+rbm.hybridize()
+trainer = mx.gluon.Trainer(
+    get_non_auxiliary_params(rbm),
+    'sgd', {'learning_rate': args.learning_rate, 'momentum': args.momentum})
+for epoch in range(args.num_epoch):
+    # Update parameters
+    for batch, _ in train_data:
+        batch = batch.as_in_context(ctx).flatten()
+        with mx.autograd.record():
+            out = rbm(batch)
+        out[0].backward()
+        trainer.step(batch.shape[0])
+    mx.nd.waitall() # To restrict memory usage
+
+    # Monitor the performace of the model
+    params = get_non_auxiliary_params(rbm)
+    param_visible_layer_bias = params['rbm_visible_layer_bias'].data(ctx=ctx)
+    param_hidden_layer_bias = params['rbm_hidden_layer_bias'].data(ctx=ctx)
+    param_interaction_weight = params['rbm_interaction_weight'].data(ctx=ctx)
+    test_log_likelihood, _ = estimate_log_likelihood(
+            param_visible_layer_bias, param_hidden_layer_bias, param_interaction_weight,
+            args.ais_batch_size, args.ais_num_batch, args.ais_intermediate_steps, args.ais_burn_in_steps, test_data, ctx)
+    train_log_likelihood, _ = estimate_log_likelihood(
+            param_visible_layer_bias, param_hidden_layer_bias, param_interaction_weight,
+            args.ais_batch_size, args.ais_num_batch, args.ais_intermediate_steps, args.ais_burn_in_steps, train_data, ctx)
+    print("Epoch %d completed with test log-likelihood %f and train log-likelihood %f" % (epoch, test_log_likelihood, train_log_likelihood))
+
+
+### Show some samples.
+
+# Each sample is obtained by 3000 steps of Gibbs sampling starting from a real sample.
+# Starting from the real data is just for convenience of implmentation.
+# There must be no correlation between the initial states and the resulting samples.
```

Comment:
Why irrelevant? I chose starting from real images insted of random images, but starting from a random images is also a common practice. Mentioning an alternative choice would be helpful here. No correlation between the initial state and the the steady state is also a general property of a Markov chain.

Reply:
Precisely, starting from real images is a common practice, whereas starting from random states is not common. 

Let's see if the following arguments make sense: The negative examples in the training loop all start from real images. The training loop may never see some true negatives from uniform random choices, because the number of visited states is far fewer than the 2^130 possibilities. As a result, swapping the sampling seed with a uniform random start may have unforeseeable outcomes.

As a matter of fact, I did the experiment with random pixel values sampled from an iid beta distribution. More than half of the images did not converge to images of digits.

In the paper I referenced, the authors trained an VAE using real images. However, visualization of the latent states reveals large areas that are not visited by the real images. Sampling from them would likely generate unreal digits. I believe this is a common issue with generative models.

Reply:
I'm not sure that the result in the paper you mentioned has how much generality for various generative models (Anyway the paper itself looks interesting). I think the trained RBM hardly define a reducible Gibbs chain. If it does, training an RBM using Gibbs sampling could be very difficult problem. As we increase the running time of the Gibbs sampling in the training, the quality of the resulting model gets better. This is a common observation, but I think it is also unexpected if the chain is reducible. Thus the Gibbs chain may be irreducible or has very limited number of isolated states. Then the initial state and the steady state of the chain must have no direct relation. If they have some relation, that means the running time of the chain is too short. 

The relaxation to the steady state is exponential except some special cases. To reach the steady state, the chain does not visit each possible image in the uniform probability. Also it does not need to visit a large fraction of the all possible images. The relaxation is exponential but the constant factor could matter in reality. So we start the chain from the real images. Starting from real images is just a trick to reduce the long relaxation time of the random initial state. Anyway it is a matter of the constant factor. It cannot beat the exponential nature of the relaxation. If you get images far from digits in the samples obtained from a *long* Gibbs chain starting from random images, they must finally appear in the samples starting from the real images. So the problem is just how long we can run chain.

From the same argument, I think comparing the initial real images and the resulting images is also not appropriate. It could be misleading. If you want to see the quality of the trained RBM, comparing some random samples from the test dataset and the resulting images of the model would be an appropriate way.

By the way, there are many papers with good RBM samples starting from random states. For example, the AIS paper you linked also shows examples of the RBM samples starting from randomly initialized MNIST images.

Reply:
I think that the probability distribution used to generate the random initial state could also matter. I'm not sure that you casted the random numbers from the beta distribution to binary units. If not, I guess it could make the relaxation time longer because much larger fraction of the state space has very low probability and starting from such states needs longer time to reach the steady state. Basically this RBM assumes the state space of binary units.

Reply:
Ok. I am convinced. The speed of convergence does not depend on the total number of states, but rather how fast any state transitions to any other state. It would be nice if there are analyses applied to RBMs in particular.

Recommend to approve unaltered.

Reply:
@yifeim are the previous batch of reviews still relevant?

Reply:
To answer your question, I was using grayscale on the image pixel values. I guess you were taking float numbers to initialize but then use binary values on the samples?

Reply:
@szha The comments are not necessary at this point. We could publish the original submission to save everyone's time.

Reply:
@asitstands Actually, what do you think about including the png that compares the original and the sampled - so that people can understand the differences and that the posteriors truly mixed?

Reply:
@yifeim Yes I mean the initialization with grayscale or black/white. The real data is grayscale, but anyway the model assumes binary pixels. This discrepancy can be ignored in many case but sometimes it makes difference. Anyway I think it is not importance in this stage.

I didn't think about demonstrating the mixing. It would be meaningful. Ok, I'll update the code to include the starting images as you suggested. That could show the quality of the trained model and also show the mixing.

Reply:
Sounds good to me and thanks for the efforts!